### PR TITLE
machine/atsamd51: pin method cleanup

### PR DESCRIPTION
**Pull request by Infinoid**
Clean up the handling of register groups.
These pin methods apply to the whole `atsamd51` family; move them over to the family-common source file.

### CODE REVIEW

**Comment by Infinoid** [comment]
I did my best to make sure these changes are correct, but I don't have a good way to verify that all of the pin modes work.  I've verified that blinking LEDs still works.  Is a battery of GPIO tests available that we can run on this?

For review purposes, it's better to look at each commit individually.

Commit 1 updates the methods in place, to improve the way register groups are handled.  This is a cleaner way to do it, and also scales nicely to the larger 120/128 pin chips which have more than 2 groups of pins.
Commit 2 is a simple cut/paste to move the methods into `machine_samd51.go`.

The chip-specific file is now mostly empty.

**Comment by Infinoid** [comment]
Is there a good way to run standard tools like `go vet` on tinygo code?  I get errors like:
```
$ go vet ./...
buffer.go:4:2: unknown import path "runtime/volatile": cannot find module providing package runtime/volatile
```

**Comment by jadr2ddude** [comment]
Not at the monent. I have an idea for a hack to do this but I haven't had a chance to try it.

File: src/machine/machine_atsamd51.go [hunk 1]
```diff
@@ -113,24 +113,21 @@ const (
 // Return the register and mask to enable a given GPIO pin. This can be used to
 // implement bit-banged drivers.
 func (p Pin) PortMaskSet() (*uint32, uint32) {
-	group := uint8(p) >> 5
-	pin_in_group := uint8(p) & 0x1f
+	group, pin_in_group := p.getPinGrouping()
```
Comment (aykevl) [inline src/machine/machine_atsamd51.go]: Why not just implement `PortMaskSet` and `PortMaskClear` and use that below in all pin manipulations?

Comment (Infinoid) [reply src/machine/machine_atsamd51.go]: `PortMaskSet` and `PortMaskClear` point to the `OUTSET` and `OUTCLR` registers.  But most of the Pin methods are using other registers, like `IN`, `DIRSET`, `DIRCLR`, `PMUX`, `PINCFG`.

Maybe I'm missing the point.  Could you go into more detail?  Or mock up some code that demonstrates your idea?

**Comment by aykevl** [comment]
@Infinoid I think this PR goes in a good direction but it's starting to get a bit big. Please don't add any more commits before it's merged, it only makes reviewing harder to do.
For example, the `Toggle()` is really a new functionality and should ideally have been done in a separate PR (but it looks good to me).

@deadprogram can you take a look at this PR, especially the first few commits? They look good to me, but you have more experience with these chips.

File: src/machine/machine_atsamd51.go [hunk 2]
```diff
@@ -141,6 +141,13 @@ func (p Pin) Get() bool {
 	return (sam.PORT.GROUP[group].IN.Get()>>pin_in_group)&1 > 0
 }
 
+// Invert the current value of a GPIO pin.
```
Comment (aykevl) [inline src/machine/machine_atsamd51.go]: The Go convention is to start the comment with the function name itself. I would suggest:

```go
// Toggle switches an output pin from low to high or from high to low.
```

Comment (Infinoid) [reply src/machine/machine_atsamd51.go]: Fixed.  Good catch.

File: src/machine/machine_atsamd51.go (hunk 1 above)
Comment (aykevl) [reply src/machine/machine_atsamd51.go]: Hmm, on second thought it'll probably only complicate things. Nevermind.

**Comment by deadprogram** [comment]
I agree with @aykevl that the `Toggle()` should have probably been a separate PR.

Tested this PR with GPIO, SPI, and I2C on my ItsyBitsy-M4 and worked as expected.

I am going to squash into a single commit, I'd suggest next time please slightly more atomic PR if possible. I know that is not always able to happen, no big deal, just trying to keep it all as clean as possible.

Now squash/merging, thank you very much for another contribution.

**Pull request merged**